### ISSUE TITLE
Pip install into virtualenv doesn't work (migrated from Trac #1857)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3405]
Original ticket http://projects.scipy.org/numpy/ticket/1857
Reported 2011-06-03 by atmention:rgommers, assigned to atmention:rgommers.

From Matthew Brett:

I've appended the result of a fresh python3.2 virtualenv and a "pip
install numpy".

```
   RefactoringTool:
/home/mb312/.virtualenvs/bare-32/build/numpy/build/py3k/numpy/core/defchararray.py
   Running from numpy source directory.Traceback (most recent call last):
     File "<string>", line 14, in <module>
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/setup.py",
line 211, in <module>
       setup_package()
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/setup.py",
line 204, in setup_package
       configuration=configuration )
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/build/py3k/numpy/distutils/core.py",
line 152, in setup
       config = configuration()
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/setup.py",
line 151, in configuration
       config.add_subpackage('numpy')
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/build/py3k/numpy/distutils/misc_util.py",
line 972, in add_subpackage
       caller_level = 2)
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/build/py3k/numpy/distutils/misc_util.py",
line 941, in get_subpackage
       caller_level = caller_level + 1)
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/build/py3k/numpy/distutils/misc_util.py",
line 878, in _get_configuration_from_setup_py
       config = setup_module.configuration(*args)
     File "numpy/setup.py", line 5, in configuration
       config = Configuration('numpy',parent_package,top_path)
     File "/home/mb312/.virtualenvs/bare-32/build/numpy/build/py3k/numpy/distutils/misc_util.py",
line 713, in __init__
       raise ValueError("%r is not a directory" % (package_path,))
   ValueError: 'build/py3k/numpy' is not a directory
```

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-06-03

Normal install: on the first pass through {{{Configuration.**init**}}} the top_path is set to numpy/build/py3k. This happens to be the right path, but it is set by accident. What happens is that the path is determined by
    caller_file = eval('**file**', frame.f_globals, frame.f_locals)
    d = os.path.dirname(os.path.abspath(caller_file))
Now, because setup.py is in the current dir, the f_globals has the following:
    (Pdb) frame.f_globals['**file**']
    'setup.py'
    (Pdb) os.path.abspath(caller_file)
    '/Users/rgommers/Code/numpy/build/py3k/setup.py'
    (Pdb) os.path.dirname(os.path.abspath(caller_file))
    '/Users/rgommers/Code/numpy/build/py3k'
There is no such file, the actual file is at '/Users/rgommers/Code/numpy/setup.py'. The reason this works is that in the main setup.py we have
    src_path = os.path.join(local_path, 'build', 'py3k')
    os.chdir(src_path)
    sys.path.insert(0, src_path)

The virtualenv install produces:
    (Pdb) frame.f_globals['**file**']
    '......./venvs/py32/build/numpy/setup.py'
Since this is an absolute path we get:
    (Pdb) os.path.dirname(os.path.abspath(caller_file))
    '/Users/rgommers/Code/venvs/py32/build/numpy'
This is the correct path looking at what the path determination is trying to do, but it doesn't work. We need "build/py3k/" appended. 

It looks like the bug is in {{{get_path_from_frame()}}}, in particular the abspath call doesn't make sense, it actually wants the current dir.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-06-03

An easy solution would be to just add something like 
        if sys.version_info[0] == 3:
            d = os.path.join(d, 'build', 'py3k')
but we can't do this because it will break other packages that use numpy.distutils.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-06-03

To get past the first error we can change the first part of {{{get_path_from_frame}}} to:
        try:
            caller_file = eval('**file**', frame.f_globals, frame.f_locals)
            d = os.path.dirname(os.path.abspath(caller_file))
            # Ugly hack to support PIP install in a virtualenv.
            if d.endswith('build/numpy'):
                d = d + 'build/numpy'

Then we run into what looks like a pip issue:
    running egg_info

```
Running from numpy source directory.error: error in 'egg_base' option: 'pip-egg-info' does not exist or is not a directory

----------------------------------------
Command python setup.py egg_info failed with error code 1
Storing complete log in /Users/rgommers/.pip/pip.log
```

this comes from {{{run_egg_info}}} in pip/req.py

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-06-03

In pip/req.py something like "python setup.py egg_info" is run in a subprocess. Before that it explicitly sets **file**, which explains the difference in values we get returned from {{{frame.f_globals['**file**']}}}.

The offending code in req.py is
        def setup_py(self):
            return os.path.join(self.source_dir, 'setup.py')
    ...
        ## FIXME: This is a lame hack, entirely for PasteScript which has
        ## a self-provided entry point that causes this awkwardness
        _run_setup_py = """
        __file__ = **SETUP_PY**
    ...
        script = script.replace('**SETUP_PY**', repr(self.setup_py))

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-06-03

So in summary:
1. There's a bug in {{{Configuration.**init**}}} where {{{dirname(abspath(nonexistingfile))}}} gives curdir.
2. The bug in (1) happens to work out well in combination with the {{{os.chdir}}} call in the main setup.py.
3. Once we correct for this so that the bug from (1) does not affect install under pip + virtualenv, we run into another issue due to both numpy.distutils and pip messing with {{{**file**}}} in ugly and incompatible ways.

I think I'll give up for now.

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-07-20

I wonder if the answer to this wouldn't be another "lame hack" in pip itself to support the numpy case?  Not that that would be ideal.  Has this been brought up with the pip developers at all for ideas?

Sorry to resurrect this issue, but it's been bugging me for a while.  I'd be willing to take a stab at it if necessary.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-07-20

I haven't tried to bring that up, and it would be a bit ugly. But if it works, why not. If you want to bring this up with them go for it. Perhaps they see a solution.

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-07-21

I played with this a bit today and made some progress with the following small diff to the main setup.py:

```
#!diff
diff --git a/setup.py b/setup.py
index 542ca7b..531364a 100755
--- a/setup.py
+++ b/setup.py
atmention:atmention: -166,6 +166,12 atmention:atmention: def setup_package():
         site_cfg = os.path.join(local_path, 'site.cfg')
         if os.path.isfile(site_cfg):
             shutil.copy(site_cfg, src_path)
+        if os.path.exists('pip-egg-info'):
+            # This means we're being installed by pip, and have to do some
+            # hackery to get this to work in Python 3
+            global __file__
+            __file__ = os.path.join(os.curdir, os.path.basename(__file__))
+            os.mkdir(os.path.join(src_path, 'pip-egg-info'))
```

This gets me a good way into the build, before it finally fails here:

```
Traceback (most recent call last):

  File "<string>", line 14, in <module>

  File "/tmp/pip-4mgbhb-build/setup.py", line 207, in <module>

    setup_package()

  File "/tmp/pip-4mgbhb-build/setup.py", line 200, in setup_package

    configuration=configuration )

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/core.py", line 186, in setup

    return old_setup(**new_attr)

  File "/home/embray/lib/python3.2/distutils/core.py", line 149, in setup

    dist.run_commands()

  File "/home/embray/lib/python3.2/distutils/dist.py", line 919, in run_commands

    self.run_command(cmd)

  File "/home/embray/lib/python3.2/distutils/dist.py", line 938, in run_command

    cmd_obj.run()

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/egg_info.py", line 8, in run

    self.run_command("build_src")

  File "/home/embray/lib/python3.2/distutils/cmd.py", line 315, in run_command

    self.distribution.run_command(command)

  File "/home/embray/lib/python3.2/distutils/dist.py", line 938, in run_command

    cmd_obj.run()

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/build_src.py", line 152, in run

    self.build_sources()

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/build_src.py", line 169, in build_sources

    self.build_extension_sources(ext)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/build_src.py", line 328, in build_extension_sources

    sources = self.generate_sources(sources, ext)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/build_src.py", line 385, in generate_sources

    source = func(extension, build_dir)

  File "numpy/core/setup.py", line 410, in generate_config_h

    moredefs, ignored = cocache.check_types(config_cmd, ext, build_dir)

  File "numpy/core/setup.py", line 41, in check_types

    out = check_types(*a, **kw)

  File "numpy/core/setup.py", line 305, in check_types

    res = config_cmd.check_type_size(complex_def, expected=2*expected[type])

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/config.py", line 252, in check_type_size

    headers, include_dirs, 'c')

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/config.py", line 99, in _compile

    (body, headers, include_dirs, lang))

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/command/config.py", line 89, in _wrap_method

    ret = mth(*((self,)+args))

  File "/home/embray/lib/python3.2/distutils/command/config.py", line 135, in _compile

    self.compiler.compile([src], include_dirs=include_dirs)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/ccompiler.py", line 26, in <lambda>

    m = lambda self, *args, **kw: func(self, *args, **kw)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/ccompiler.py", line 199, in CCompiler_compile

    self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/ccompiler.py", line 26, in <lambda>

    m = lambda self, *args, **kw: func(self, *args, **kw)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/unixccompiler.py", line 35, in UnixCCompiler__compile

    extra_postargs, display = display)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/ccompiler.py", line 26, in <lambda>

    m = lambda self, *args, **kw: func(self, *args, **kw)

  File "/tmp/pip-4mgbhb-build/build/py3k/numpy/distutils/ccompiler.py", line 61, in CCompiler_spawn

    print(o)

UnicodeEncodeError: 'ascii' codec can't encode characters in position 50-52: ordinal not in range(128)
```

Not sure why that would happen.  When I do a normal `./setup.py build` it looks like gcc is outputting some warnings that contain non-ascii characters.  But why that's a problem when installing with pip but not otherwise isn't clear.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-07-24

Don't you want 
    **file** = os.path.join(src_path, os.path.basename(**file**))
instead of
    **file** = os.path.join(os.curdir, os.path.basename(**file**))
?

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-07-25

I imagine that could also work.  But when I ran the setup.py without pip in PDB it showed that **file** was just `./setup.py`, not an absolute path, when it gets to `get_path_from_frame()`, so that line was just restoring **file** to what it would have been if not for pip.

From that point on the install seems to proceed normally until it gets to this strange encoding issue when calling gcc.  I know that pip captures stdout and reoutputs it, so my guess is that it's a separate problem in pip.

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-07-25

The encoding error I got above is due to this issue in the subprocess module: http://bugs.python.org/issue6135

pip uses subprocess.Popen to run the setup.py, when then gets its stdout encoding set to ASCII.  I'm not sure yet what the workaround for that would be, but it should probably be on pip's end.

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-07-25

Well, I commented out the `print(o)` that was causing problems at line 61 of numpy/distutils/ccompiler.py, and updated my patch to the setup.py to read:

```
#!patch
diff --git a/setup.py b/setup.py
index 542ca7b..ef70d16 100755
--- a/setup.py
+++ b/setup.py
atmention:atmention: -166,6 +166,16 atmention:atmention: def setup_package():
         site_cfg = os.path.join(local_path, 'site.cfg')
         if os.path.isfile(site_cfg):
             shutil.copy(site_cfg, src_path)
+        if os.path.exists('pip-egg-info'):
+            # This means we're being installed by pip, and have to do some
+            # hackery to get this to work in Python 3
+            # This won't work if pip was run with the --editable option, but
+            # numpy won't work in that context anyways
+            global __file__
+            __file__ = os.path.join(os.curdir, os.path.basename(__file__))
+            if '--egg-base' in sys.argv:
+                idx = sys.argv.index('--egg-base')
+                sys.argv[idx + 1] = os.path.join(local_path, 'pip-egg-info')
```

and with that the install succeeded.  So I would recommend adding this patch, or something like it to numpy.  In the meantime I'll get in touch with the pip developers about this encoding issue.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-07-25

That sounds promising - I'll try to test and review this soon.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-02

Can you explain why this doesn't work with -e / --editable? This is the recommended way for install from local sources or local git repo. If that doesn't work it's kind of a problem, plus it makes it difficult to test the change. What did you do, push to your github account and then install it from there?

pip not having a normal interface for installing from local files / checkout is completely ridiculous by the way.

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-08-02

I see what you're saying; I didn't realize that numpy ''should'' work with --editable.

The reason this patch would ''not'' work with --editable is that it relies on `--egg-base pip-egg-info` being in `sys.argv` to determine that the egg_info argument is being run by pip.  However, that argument is not present when installing with --editable.  I don't think we can rely on looking up the frame stack since pip runs the setup.py in a subprocess.

However, I tried removing the test for pip-egg-info and instead just fixing up `__file__` regardless of whether we're being installed by pip, and that seems to work just fine, including with --editable.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-02

This works with my local checkout and doing "pip install -e .". In that way {{{--egg-base}}} is not in sys.argv though. How did you test that?

Can you also explain what these lines do?
            if '--egg-base' in sys.argv:
                idx = sys.argv.index('--egg-base')
                sys.argv[idx + 1] = os.path.join(local_path, 'pip-egg-info')

I also didn't find a reliable way to detect whether the install is being driven by pip unfortunately.

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-08-03

Really?  Using the last patch I posted `pip install -e .` does not work--the install crashes as the usual place, with "'build/py3k/numpy' is not a directory".  Relying on 'pip-egg-info' was the only way it could determine if it was being run by pip.

The lines you asked about fix up the path for the pip-egg-info directory.  By default the path passed to --egg-base is relative to the current working directory.  So when the setup.py changes directories to build/py3k/numpy, the pip-egg-info directory gets created at build/py3k/numpy/pip-egg-info.  But after the `setup.py egg_info` command is run, pip goes looking for that directory and doesn't find it.  So what those lines do is it just fixes the path for pip-egg-info so that it's absolute, and in the location where pip expects to find it.  Does that make sense?

For what it's worth, this is what my setup.py looks like now:

```
#!diff
diff --git a/setup.py b/setup.py
index 542ca7b..ec05e30 100755
--- a/setup.py
+++ b/setup.py
atmention:atmention: -166,6 +166,14 atmention:atmention: def setup_package():
         site_cfg = os.path.join(local_path, 'site.cfg')
         if os.path.isfile(site_cfg):
             shutil.copy(site_cfg, src_path)
+        # pip messes with __file__ which can cause problems when installing in
+        # Python3; restore __file__ to what it would have been otherwise
+        global __file__
+        __file__ = os.path.join(os.curdir, os.path.basename(__file__))
+        if '--egg-base' in sys.argv:
+            idx = sys.argv.index('--egg-base')
+            if sys.argv[idx + 1] == 'pip-egg-info':
+                sys.argv[idx + 1] = os.path.join(local_path, 'pip-egg-info')

     old_path = os.getcwd()
     os.chdir(src_path)
```

It now fixes `__file__` regardless of whether we're in pip--if we're not in pip then it makes no difference.  If `--egg-base pip-egg-info` is in sys.path then it makes that fix to the pip-egg-info path.

Then there's just the one issue with pip choking on non-ascii output from the build.  I created a [https://github.com/pypa/pip/issues/325 bug report] for that, but it hasn't been noticed yet, so I might go poke them on the mailing list or something.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-03

I meant after always changing {{{**file**}}} it works. Looks like this for me:
    #!diff
    diff --git a/numpy/distutils/ccompiler.py b/numpy/distutils/ccompiler.py
    index 14213dc..e3b88af 100644
    --- a/numpy/distutils/ccompiler.py
    +++ b/numpy/distutils/ccompiler.py
    atmention:atmention: -58,7 +58,11 atmention:atmention: def CCompiler_spawn(self, cmd, display=None):
         if s:
             if is_sequence(cmd):
                 cmd = ' '.join(list(cmd))
    -        print(o)
    +        try:
    +            print(o)
    +        except UnicodeError:
    +            # When installing through pip, `o` can contain non-ascii chars
    +            pass
             if re.search('Too many open files', o):
                 msg = '\nTry rerunning setup command until build succeeds.'
             else:
    diff --git a/setup.py b/setup.py
    index 542ca7b..b0b7a97 100755
    --- a/setup.py
    +++ b/setup.py
    atmention:atmention: -167,6 +167,18 atmention:atmention: def setup_package():
             if os.path.isfile(site_cfg):
                 shutil.copy(site_cfg, src_path)

```
+        # Ugly hack to make pip work with Python 3, see #1857.
+        # Explanation: pip messes with __file__ which interacts badly with the
+        # change in directory due to the 2to3 conversion.  Therefore we restore
+        # __file__ to what it would have been otherwise.
+        global __file__
+        __file__ = os.path.join(os.curdir, os.path.basename(__file__))
+        if '--egg-base' in sys.argv:
+            # Change pip-egg-info entry to absolute path, so pip can find it
+            # after changing directory.
+            idx = sys.argv.index('--egg-base')
+            sys.argv[idx + 1] = os.path.join(local_path, 'pip-egg-info')
+
     old_path = os.getcwd()
     os.chdir(src_path)
     sys.path.insert(0, src_path)
```

Works for me with -e. Haven't tested without -e; is there a better way to do that than pushing to my github account and letting pip install from there with some magic incantation?

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-08-03

Replying to [comment:18 rgommers]:

> Works for me with -e. Haven't tested without -e; is there a better way to do that than pushing to my github account and letting pip install from there with some magic incantation?

Yes, you can just run `pip install .` on your source checkout.  It will copy the source to a temp dir (I think), create an egg, and then install the egg.

Your patch looks almost exactly like mine now.  They only difference is that it should only fix the -egg-base argument ''if'' its prior value is 'pip-egg-info'.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-03

Ah, that does work. I tried that before but it seemed to hang. pip is just thinking for a long time without telling me what it's doing.

Okay, works for both install methods, and normal {{{python setup.py install}}} works too.  So job done I think.  How do you want me to credit you, is it "E.M. Bray"?

**Comment by thouis** [comment]
Comment in Trac by trac user embray, 2011-08-03

Cool!  "Erik Bray" is fine; thanks.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-08-13

Should be fixed in commit:dba98cc1 now.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-04-30
